### DISCUSSION TITLE
Using import in cue file inside golang project

### DISCUSSION THREAD

**Discussion by roylisto** [cue-lang/cue#1139] [Q&A]
Hi, I want to use my cue files inside golang code. 
My cue files using `import` to use another field inside other cue files.

How to load the cue file with dependencies and how the folder structure should be? is it also need a cue.mod?

**Comment by mboogerd** [comment]
Hi, I've been reverse engineering the codebase last week for this exact question (I wanted the openapi code gen from cue definition, which works unfortunately only for simple cases). Here's my attempt that seems to work to get the validated cue definition though:

```go
import (
	"cuelang.org/go/cue"
	"cuelang.org/go/cue/build"
	"cuelang.org/go/cue/load"
)

func loadCueInstanceFromPath(path string) *cue.Instance {

	load := load.Instances([]string{path}, nil)
	// todo: safe access
	build_instance := load[0]

	// new way of doing things, but openapi does not take `Value` yet
	// ctx := cuecontext.New()
	// cue_value := ctx.BuildInstance(build_instance)

	// deprecated, but OpenAPI generator can take this object
	cue_instances := cue.Build([]*build.Instance{build_instance})

	// todo: safe access
	return cue_instances[0]
}
```

Just pass a path to your cue file to this function and it should work. If the file's imports resolve using the cue CLI, they should resolve also with this function. At least that was the case for me. 

_Note_: no safety built in:
- dereferencing zero index of array
- The function will load an instance with some error set if the loaded cue document is invalid IIRC.

**Comment by myitcv** [comment]
@roylisto are you looking to embed the CUE fields in your Go program? If so, take a look at https://github.com/cue-lang/cue/discussions/1117

Otherwise, an approach like @mboogerd's will work. A `cue.mod` will be required if you are importing non-builtins. 

If you have an example you can share, that might help us to be a bit more precise.

**Reply by roylisto** [reply]
Hi, @myitcv I am also want to know how to embed cue files inside golang code so I can distribute it as a cli app. Is it better to create another discussion for that?

**Reply by verdverm** [reply]
#1117 has some example code for that

**Reply by roylisto** [reply]
> #1117 has some example code for that

so I don't have to use //go:embed to embed cue file?

**Reply by myitcv** [reply]
I'm not sure I follow your question, @roylisto. If you can present a more complete idea of your situation that will help.

**Comment by roylisto** [comment]
```
package metadata

import "example.com/util"

#Metadata: {
  name: util.#StringValidation
  description?: string // optional
  ... // allow any fields
}

#Metadata

```

Here is my example, will `import` works? if yes, how should I use `import` inside golang code? is it by package name?
